### ISSUE TITLE
Functionality of Sparse/PassThrough encoders & mux/demux & vision

### ISSUE THREAD

**Issue by breznak** [numenta/nupic#2065]
I'd like to review functionality of `PassThrough` & `SparsePassThrough` encoders, and discuss some enhancements. 

# PassThrough: 
* with `force=True` just passes input to nupic (with #2064 1D,2D,... input can be used)
* with `w` set only checks (raises err) if sparsity,w in input data does not meet given `w`
 * I thought before it used to/should _enforce_ that sparsity - meaning ON bits would be randomly set (added/removed) to achieve given `w` in each input. (or a new param `enforceWbitsON`?)

# SparsePassThrough:
* accepts the input array given as sparse (defined by indices) for PassThrough
 * can this functionality be merged into a single encoder with a param `isSparseInput` ?
* here `w` implements a "demux" (multiplier) for w>1 each bit is represented by w-bits (w=3: `01`->`000111`)
 * I'd like to implement this functionality (and meaning) of `w` also in PassThrough
 * the "check sparsity-w" could be renamed to `checkWbitsON` :tongue: 
 * the code logic would only move from SparsePassTh to PassTh
 * could be useful for vision
* in future, consider `w<1` as a "mux" (compression); eg (w=3: `000111`->`01`, but what for `010`->? (0, ...?))

# Vision
* I plan to use the "demux", 2D for PassThrough encoder in vision, should I enhance the PassTh, or extend it to a new encoder? (I think so far the above features would be generally useful)
* use-case: I will be encoding sparse 2D bitmaps and plan to use (Sparse)PassThrough encoder for that
 * eg :  
```
0 1
1 0
```
* cont..
  * but sometimes I will have (quantized) values (eg 4 values), eg : 
```
0 1
3 0
```
* cont..
 * I can discretize them (and lose info), or try to "encode" in 2D pass-through; I was thinking like (for [0,3] values):
   * "demux" where `w=4: 0->0000, 1->1000, ..., 3->1111`; that in 2D, and "ordering" does not matter (1000=0100=0001=...)
   * this "colorPassThrough" encoder would probably need its separate class

A lot of questions, what's your thoughts? :wink:

**Comment by breznak** [comment]
CC @chetan51 @scottpurdy @subutai

**Comment by scottpurdy** [comment]
These pass through encoders are provided so that you can handle encoding yourself. I wouldn't want to add different encoding options to them since the whole point is that the caller should be encoding however they like. Even the `w` option seems unnecessary to me. If you are trying to do vision, I would *strongly* suggest checking out the existing code that has a tremendous amount of functionality based on extensive experimentation and expertise.

- [ImageSensor Region](https://github.com/numenta/nupic/blob/master/nupic/regions/ImageSensor.py)
- [Example usage](https://github.com/numenta/nupic.research/blob/master/image_test/run_mnist_experiment.py)

It is totally fine if you want to build your own image encoders but we probably wouldn't include them in NuPIC unless you convince us that they are superior for some large class of problems. There may be some ways to simplify our current encoder or break it into different classes though, but this would need a really thoughtful, detailed effort.

**Comment by breznak** [comment]
Yep, I'm aware of `ImageSensor` but it does not meet my requirements - it's more like a "framework" for loading/presenting/modifying images, as I understand it (and it's quite complex/big); I basically want a "smarter" extension to PassThrough for presenting a sparse bitmap... 

It would be nice though, if I can reuse as much as possible from PassThrough encoders, so what do you mean about the suggested changes in the 2 classes? I think some of the functionality is not very useful and can be dropped; And what about the `w` (multiplier)? Remove completely, or move to PassTh, so both classes use the same functionality?

**Comment by scottpurdy** [comment]
I'd say remove if anything but probably just leave alone for now.
